caps/types.go

```diff
@@ -55,11 +58,26 @@ func (t *Type) Validate(c *Capability) error {
 	if t != c.Type {
 		return fmt.Errorf("capability is not of type %q", t)
 	}
-	// While we don't have any support for type-specific attribute schema,
-	// let's ensure that attributes are totally empty. This will make tests
-	// show that this code is actually being used.
-	if c.Attrs != nil && len(c.Attrs) != 0 {
-		return fmt.Errorf("attributes must be empty for now")
+	// Check that all required attributes are present
+	for _, attr := range t.RequiredAttrs {
+		if _, ok := c.Attrs[attr]; !ok {
+			return fmt.Errorf("capability lacks required attribute %q", attr)
+		}
+	}
+	// Look for any unexpected attributes
```

Comment:
Can you explain this more please? I've made it precisely so that we have forward compatibility. Otherwise anyone can ship a product with a "foo" type that has a custom, unexpected "bar" attribute that we want to use but now cannot because shipping products and snaps have already used it.

Down the line I have an idea for type versions so that we can release, say, "bool-file" v2 with more attributes and applications that want to see that can use the new attributes (perhaps with new usage semantics as well, that are not strictly encoded in attributes but this is unrelated).

When we ship a type we ship it with fixed set of attributes. Each time we want to change what they are, what they mean, we bump the version and ship another capability (version) that represents the same resource. This gives us future-compatibility by letting us evolve types over time.

Obviously you had something else on your mind so please explain what you mean again and I'll adjust the code.